src/pydvl/influence/frameworks/torch_differentiable.py

```diff
@@ -505,3 +524,119 @@ def solve_arnoldi(
+    """
+    Calculates a low-rank approximation of the Hessian matrix of the model's loss function using the implicitly
+    restarted Lanczos algorithm.
+
+
+    :param hessian_vp: A function that takes a vector and returns the product of the Hessian of the loss function
+    :param matrix_shape: The shape of the matrix, represented by hessian vector product.
+    :param hessian_perturbation: Optional regularization parameter added to the Hessian-vector product
+                                 for numerical stability.
+    :param rank_estimate: The number of eigenvalues and corresponding eigenvectors to compute.
+                          Represents the desired rank of the Hessian approximation.
+    :param krylov_dimension: The number of Krylov vectors to use for the Lanczos method.
+                             If not provided, it defaults to $min(model.num_parameters, max(2*rank_estimate + 1, 20))$.
+    :param x0: An optional initial vector to use in the Lanczos algorithm.
+               If not provided, a random initial vector is used.
+    :param tol: The stopping criteria for the Lanczos algorithm, which stops when the difference
+                in the approximated eigenvalue is less than `tol`. Defaults to 1e-6.
+    :param max_iter: The maximum number of iterations for the Lanczos method. If not provided, it defaults to
+                     $10*model.num_parameters$
+    :param device: The device to use for executing the hessian vector product.
+    :param eigen_computation_on_gpu: If True, tries to execute the eigen pair approximation on the provided
+                                     device via cupy implementation.
+                                     Make sure, that either your model is small enough or you use a
+                                     small rank_estimate to fit your device's memory.
+                                     If False, the eigen pair approximation is executed on the CPU by scipy wrapper to
+                                     ARPACK.
+    :return: A `LowRankProductRepresentation` instance that contains the top (up until rank_estimate) eigenvalues
+             and corresponding eigenvectors of the Hessian.
+    """
+
+    if eigen_computation_on_gpu:
+        try:
+            import cupy as cp
+            from cupyx.scipy.sparse.linalg import LinearOperator, eigsh
+            from torch.utils.dlpack import from_dlpack, to_dlpack
+        except ImportError as e:
+            raise ImportError(
+                f"Try to install missing dependencies or set eigen_computation_on_gpu to False: {e}"
+            )
+
+        if device is None:
+            raise ValueError(
+                "Without setting an explicit device, cupy is not supported"
+            )
+
+        def mv_cupy(x):
+            x = from_dlpack(x.toDlpack())
+            y = hessian_vp(x) + hessian_perturbation * x
+            return cp.from_dlpack(to_dlpack(y))
+
+        eigen_vals, eigen_vecs = eigsh(
```

Comment:
There is a lot of repetition and return points in this function. I think it should be simplified by making a single call to `eigsh` and defining the different variables and functions in the if and else branches, along with an extra function to convert the result to torch tensors.

Reply:
Agree, done